crates/shared/src/event_handling.rs

```diff
@@ -217,31 +319,99 @@ where
+        if blocks.is_empty() {
+            return;
+        }
+        // delete forked blocks
+        self.last_handled_blocks
+            .retain(|block| block.0 < blocks.first().unwrap().0);
+        // append new canonical blocks
+        self.last_handled_blocks.extend(blocks.iter());
+        // cap number of blocks to MAX_REORG_BLOCK_COUNT
+        let start_index = self
+            .last_handled_blocks
+            .len()
+            .saturating_sub(MAX_REORG_BLOCK_COUNT as usize);
+        self.last_handled_blocks = self.last_handled_blocks[start_index..].to_vec();
+        tracing::debug!(
+            "last_handled_blocks after update: {:?} - {:?}",
+            self.last_handled_blocks.first(),
+            self.last_handled_blocks.last(),
+        );
+    }
 }
 
-/// Try to shorten the current_blocks by detecting the reorg from previous event update.
+/// Try to shorten the latest_blocks by detecting the reorg from previous event update.
 /// If no reorg can be detected (for example, when `handled_blocks` is shorter then the
-/// reorg depth) then fallback to full `current_block` as a safe measure.
+/// reorg depth) then fallback to full `latest_blocks` as a safe measure.
 fn detect_reorg_path<'a>(
     handled_blocks: &[BlockNumberHash],
-    current_blocks: &'a [BlockNumberHash],
+    latest_blocks: &'a [BlockNumberHash],
 ) -> (&'a [BlockNumberHash], bool) {
-    // in most cases, current_blocks = handled_blocks + 1 newest block
+    // in most cases, latest_blocks = handled_blocks + 1 newest block
     // therefore, is it more efficient to put the handled_blocks in outer loop,
     // so everything finishes in only two iterations.
     for handled_block in handled_blocks.iter().rev() {
-        for (i, current_block) in current_blocks.iter().enumerate().rev() {
-            if current_block == handled_block {
+        for (i, latest_block) in latest_blocks.iter().enumerate().rev() {
+            if latest_block == handled_block {
                 // found the same block in both lists, now we know the common ancestor, don't include the ancestor
                 let is_reorg = handled_block != handled_blocks.last().unwrap();
-                return (&current_blocks[i + 1..], is_reorg);
+                return (&latest_blocks[i + 1..], is_reorg);
             }
         }
     }
 
     // reorg deeper than the EventHandler history (`handled_blocks`), return full list
     let is_reorg = !handled_blocks.is_empty();
-    (current_blocks, is_reorg)
+    (latest_blocks, is_reorg)
+}
+
+/// Splits range into two disjuctive consecutive ranges, second one containing last (up to)
+/// MAX_BLOCKS_QUERIED elements, first one containing the rest (if any)
+fn split_range(
+    range: RangeInclusive<u64>,
+) -> Result<(Option<RangeInclusive<u64>>, RangeInclusive<u64>)> {
+    ensure!(!range.is_empty(), "invalid range to split");
+    ensure!(
+        MAX_BLOCKS_QUERIED > 0,
+        "MAX_BLOCKS_QUERIED must be greater than zero"
+    );
```

Comment:
1. Check for `MAX_BLOCKS_QUERIED` is more as a sanity here, so if someone changes it in the future UT would fail. I think I would be ok with removing it. I would avoid passing it as an argument as I expect it to never change.
2. Changed the code to support for empty range (we have an outside check). The `empty` is a bit unfortunate term in our case, because it indicates range goes backwards which is an error in our case rather than empty.
3. I think this is problematic because, when I want to return empty range there is no easy way as for Vec for example, where you just return vec![]. In case of RangeInclusive I would have to actually build empty range with some invalid values for example RangeInclusive::new(1,0) in order to make it empty which feels even more weird IMO.

Reply:
Since empty ranges are an error anyway what do you think about building your own range where that can't happen?

```rust
struct RangeInclusive<T: Ord> {
    start: T,
    end: T,
}

impl<T: Ord> RangeInclusive<T> {
    pub fn try_new(start: T, end: T) -> Result<Self> {
        ensure!(end >= start, "end has to be bigger than start");
        Ok(Self { start, end })
    }
    pub fn start(&self) -> T { self.start }
    pub fn end(&self) -> T { self.end }
}
```

Reply:
Neat idea!

Reply:
I think I did not express myself well.
In this particular it is invalid value but we still need to propagate it back.

Reply:
By making `split_range()` take the custom `RangeInclusive` you make sure the error (empty range) has to be handled before calling `split_range()` which is what you are currently doing [here](https://github.com/cowprotocol/services/pull/535/files#diff-60652ce9b004202bff1704ee459d23bc170eadb86f66cf5c0161a431149086c2R151-R163) in `event_handling.rs`. The difference is that you don't have to recheck `!range.is_empty()` every time you use the same range in different functions. If the optional range exists `is_some()` you know that the contained range can not be empty which appears to be an invalidate state.

Reply:
Ah yes, previously I understood that you suggested to completely remove `Option`.
Ok, added custom `RangeInclusive`, seems a bit more clear now.